lib/charms/observability_libs/v0/cert_handler.py

```diff
@@ -0,0 +1,361 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+"""## Overview.
+
+This document explains how to use the `CertHandler` class to
+create and manage TLS certificates through the `tls_certificates` interface.
+
+The goal of the CertHandler is to provide a wrapper to the `tls_certificates`
+library functions to make the charm integration smoother.
+
+## Library Usage
+
+This library should be used to create a `CertHandler` object, as per the
+following example:
+
+```python
+cert_handler = CertHandler(
+    charm=self,
+    peer_relation_name="replicas",
+    cert_subject="unit_name",  # Optional
+    key="cert-manager"  # Optional
+)
+```
+
+This library requires a peer relation to be declared in the requirer's metadata. Peer relation data
+is used for "persistent storage" of the private key and certs.
+"""
+import json
+import socket
+from typing import List, Optional, Union
+
+try:
+    from charms.tls_certificates_interface.v2.tls_certificates import (  # type: ignore
+        AllCertificatesInvalidatedEvent,
+        CertificateAvailableEvent,
+        CertificateExpiringEvent,
+        CertificateInvalidatedEvent,
+        TLSCertificatesRequiresV2,
+        generate_csr,
+        generate_private_key,
+    )
+except ImportError:
+    raise ImportError(
+        "charms.tls_certificates_interface.v2.tls_certificates is missing; please get it through charmcraft fetch-lib"
+    )
+import logging
+
+from ops.charm import CharmBase, RelationBrokenEvent
+from ops.framework import EventBase, EventSource, Object, ObjectEvents
+from ops.model import Relation
+
+logger = logging.getLogger(__name__)
+
+
+LIBID = "b5cd5cd580f3428fa5f59a8876dcbe6a"
+LIBAPI = 0
+LIBPATCH = 1
+
+
+class CertChanged(EventBase):
+    """Event raised when a cert is changed (becomes available or revoked)."""
+
+
+class CertHandlerEvents(ObjectEvents):
+    """Events for CertHandler."""
+
+    cert_changed = EventSource(CertChanged)
+
+
+class CertHandler(Object):
+    """CertHandler is used to wrap TLS Certificates management operations for charms.
+
+    CerHandler manages one single cert.
+
+    Args:
+        key: A manually-crafted, static, unique identifier used by ops to identify events.
+        It shouldn't change between one event to another.
+        peer_relation_name: Must match metadata.yaml.
+        certificates_relation_name: Must match metadata.yaml.
+        cert_subject: Custom subject. Name collisions are under the caller's responsibility.
+        extra_sans_dns: Any additional DNS names apart from FQDN.
```

Comment:
We usually place these arguments in the `__init__` docstring.

Reply:
Fixed in next commit.